### PR TITLE
Validate datasource name duplication before the save

### PR DESCRIPTION
**Pull request by acumarav**
Validate datasource name duplication before the save.
Show user friendly validation error message.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -265,6 +269,12 @@
             <version>1.3.0</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.13.2</version>
```
Comment (YaroslavTir) [inline pom.xml]: can you extract artifact version as property? 
Do we really need this library? I do not see that we are using it other projects.

File: src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java [hunk 2]
```diff
@@ -121,20 +122,56 @@ protected BaseDataSourceController(UserService userService,
...
         return result;
     }
 
+    private Optional<JsonResult> validatePersistDataSourceRequest(BindingResult bindingResult, String dataSourceName, Long dataSourceId){
+
+        if (bindingResult.hasErrors()) {
+            return Optional.of(setValidationErrors(bindingResult));
+        }
+
+        if(!dataSourceService.isDatasourceNameUnique(dataSourceName, dataSourceId)){
```
Comment (YaroslavTir) [inline src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java]: I would prefer reverse this condition, I don't really like negation in if condition:
```
if (dataSourceService.isDatasourceNameUnique(dataSourceName, dataSourceId)){
 return Optional.empty();
}
```

File: src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java [hunk 3]
```diff
@@ -350,4 +344,12 @@ public JsonResult unpublishAndDeleteOnCentral(Long dataSourceId) {
 
         return dataSourceRepository.findAllByCentralIdIsNull().collect(Collectors.toList());
     }
+
+    @Override
+    public boolean isDatasourceNameUnique(String name, Long dataSourceId) {
+
+        return dataSourceRepository.findFirstByNameIgnoreCase(name)
```
Comment (YaroslavTir) [inline src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java]: Base on the function name, one first dataSource by name will be found. Is it possible that we already have a few DS with the same name?

File: src/test/java/com/odysseusinc/arachne/datanode/controller/DataSourceControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class DataSourceControllerTest {
```
Comment (YaroslavTir) [inline src/test/java/com/odysseusinc/arachne/datanode/controller/DataSourceControllerTest.java]: well done, writing tests are cool. :)

File: pom.xml (hunk 1 above)
Comment (acumarav) [reply pom.xml]: done

File: src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java (hunk 3 above)
Comment (acumarav) [reply src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java]: I have changed this code a bit, but in general the datanode DB has unique constraint on the name column so only one result can be expected

File: pom.xml (hunk 1 above)
Comment (acumarav) [reply pom.xml]: this lib is already part of spring-boot-starter-test. I want to use newer version. But lib itself is very cool.

File: src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java (hunk 2 above)
Comment (acumarav) [reply src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java]: it is possible to introduce variable like:
bool isDatasourceWithTheSameNameAlreadyExists = ataSourceService.isDatasourceNameUnique(dataSourceName, dataSourceId);
if(isDatasourceWithTheSameNameAlreadyExists){
return Error....
}

but I do not like in offered code snipped that one validation criteria become a reason to resolve entire validation chain. It will be working only if this that check is the last one. IMHO code will not become cleaner if I apply this change.

File: src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java [hunk 5]
```diff
@@ -121,20 +122,56 @@ protected BaseDataSourceController(UserService userService,
...
+        dataSource.setId(dataSourceId);
+
+        DataSource persistedDataSource = persistDatasource.apply(user, dataSource);
         JsonResult<DataSourceDTO> result = new JsonResult<>(NO_ERROR);
-        result.setResult(conversionService.convert(optional, DataSourceDTO.class));
+        result.setResult(conversionService.convert(persistedDataSource, DataSourceDTO.class));
         return result;
     }
 
+    private Optional<JsonResult> validatePersistDataSourceRequest(BindingResult bindingResult, String dataSourceName, Long dataSourceId){
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java]: Could you parameterize method to avoid warnings about unchecked assignments?

File: src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java [hunk 6]
```diff
@@ -121,20 +122,56 @@ protected BaseDataSourceController(UserService userService,
...
+
+        if (bindingResult.hasErrors()) {
+            return Optional.of(setValidationErrors(bindingResult));
+        }
+
+        if(!dataSourceService.isDatasourceNameUnique(dataSourceName, dataSourceId)){
+            JsonResult notUniqueNameResult = new JsonResult<>(JsonResult.ErrorCode.VALIDATION_ERROR);
+            notUniqueNameResult.getValidatorErrors().put("name", Constants.DataSourceMessages.DATASOURCE_NAME_UNIQUE);
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java]: But why not to use `JsonResult.setValidatorErrors` with a `Map` containing single element as an argument?

File: src/main/java/com/odysseusinc/arachne/datanode/exception/PermissionDeniedException.java [hunk 7]
```diff
@@ -22,7 +22,7 @@
 
 package com.odysseusinc.arachne.datanode.exception;
 
-public class PermissionDeniedException extends Exception {
+public class PermissionDeniedException extends RuntimeException {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/datanode/exception/PermissionDeniedException.java]: There are lot of usage of this. Why this changes required?

File: src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java [hunk 8]
```diff
@@ -354,4 +348,12 @@ public JsonResult unpublishAndDeleteOnCentral(Long dataSourceId) {
 
         return dataSourceRepository.findAllByCentralIdIsNull().collect(Collectors.toList());
     }
+
+    @Override
+    public boolean isDatasourceNameUnique(String name, Long dataSourceId) {
+
+        return dataSourceRepository.findByName(name)
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java]: Probably `countByName` would be better and faster?

File: src/main/java/com/odysseusinc/arachne/datanode/exception/PermissionDeniedException.java (hunk 7 above)
Comment (acumarav) [reply src/main/java/com/odysseusinc/arachne/datanode/exception/PermissionDeniedException.java]: On one side - Security is checked/handled by the spring so it is overkill to force user handle it too..
and it is inconvenient to use methods with throws non-runtime exceptions in the lamdas.
So I think it will be good to mark them as Runtime to avoid necessary try {
}catch
in business logic.

File: src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java (hunk 8 above)
Comment (acumarav) [reply src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java]: count will not work here because I need to check update as well - 
so logic here find Datasource by the name and allow to save if it has the same ID:

return dataSourceRepository.findByName(name)
                .map(ds -> ds.getId().equals(dataSourceId))
                .orElse(true);

If to change to count then update will not work

Comment (wivern) [reply src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java]: But `countByNameAndNotId` works?

File: src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java (hunk 6 above)
Comment (wivern) [reply src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java]: @acumarav what you think about comment above?

File: src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java (hunk 5 above)
Comment (wivern) [reply src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java]: @acumarav could you make a parameterized method?

File: src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java (hunk 8 above)
Comment (acumarav) [reply src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java]: I do not think that it will work for the new entity case - it has Id Null and it will cause JPA exception ID cannot be null something.... So it would requre 2 method countByName and countByNameAndId
and logic like int count = (entity.id ==null)? repo.countByName(entity.name):repo.countByNameAndId(entity.name, entity.id)

but such a code does not look good for me.

File: src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java (hunk 5 above)
Comment (acumarav) [reply src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java]: done

File: src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java (hunk 6 above)
Comment (acumarav) [reply src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java]: I have changed it but j in the current codebase there are 46 usages of: .getValidatorErrors().put(
and only 2 of: .setValidatorErrors(
so I would say both ways are correct

File: src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java (hunk 8 above)
Comment (wivern) [reply src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java]: In contrast, my concern is the following:
`DataSource` is a heavy object with multiple fields and at least one relation,
why we need to restore completed object graph from database while we require just a single number?
When You don't like ternary operator you always can avoid it with strategy or with implementing a functional way.

Comment (acumarav) [reply src/main/java/com/odysseusinc/arachne/datanode/service/impl/DataSourceServiceImpl.java]: I have changed mostly because I want to stop this discussion, but I do not believe that couple of extra selects by the primary key can somehow slow down the application.
